app/models/location.rb

```diff
@@ -20,56 +20,54 @@ class Location < ActiveRecord::Base
   belongs_to :district
   belongs_to :state
 
-  after_save :update_nation_builder
+  validates :state, presence: true, unless: :zip_code
+  validates :zip_code, presence: true, unless: :state
+  validates :zip_code, allow_nil: true,
+    format: { with: /\A\d{5}[^\w]?(\d{4})?\z/ }
 
-  def update_location(address_params)
-    address_params[:zip] = nil unless ZipCode.valid_zip?(address_params[:zip])
+  ADDRESS_FIELDS = [
+    :address_1, :address_2, :city, :state_id, :zip_code, :district_id
+  ]
+  PERMITTED_PARAMS = ADDRESS_FIELDS + [:state_abbrev]
 
-    if source = address_source(address_params)
-      new_attributes = {
-        address_1:  address_params[:address] || nil,
-        city:       address_params[:city],
-        state:      source.try(:state),
-        zip_code:   source.try(:zip_code) || address_params[:zip]
-      }.compact!
+  def state_abbrev
+    (state || district).try(:abbrev)
+  end
 
-      if new_attributes.nil?
-        Airbrake.notify(error_messasage: 'Error in lookup - perhaps service is down?')
-        new_attributes = {}
-      end
-      new_attributes[:district] = source.is_a?(District) ? source : source.try(:single_district)
+  def state_abbrev=(abbrev)
+    self.state = State.find_by(abbrev: abbrev)
+  end
 
-      update_attributes(new_attributes)
-    elsif new_state = State.find_by(abbrev: address_params[:state_abbrev])
-      update_attributes(
-        address_1: address_params[:address],
-        city: address_params[:city],
-        state: new_state,
-        zip_code: address_params[:zip]
-      )
-    end
+  def fill_in_missing_attributes
+    fill_in_state
+    fill_in_district
   end
 
-  def address_source(address_params)
-    if address_params[:address]
-      District.find_by_address(address_params)
-    elsif address_params[:zip] && updated_zip?(address_params[:zip])
-      ZipCode.find_by(zip_code: address_params[:zip])
-    end
+  def fill_in_state
+    self.state_id ||= find_zip_code.try(:state_id)
   end
 
-  def state_abbrev
-    (state || district).try(:abbrev)
+  private
+
+  def fill_in_district
+    self.district ||= find_district_by_zip || find_district_by_address
   end
 
-  def state_name
-    state.try(:name)
+  def find_district_by_zip
+    find_zip_code.try(:single_district)
   end
 
-  private
+  def find_district_by_address
+    if address_1 && zip_code
+      search_params = attributes.symbolize_keys.
+        slice(:city, :state_abbrev, :zip_code).
+        merge(address: address_1)
+      District.find_by_address(search_params)
+    end
+  end
 
-  def updated_zip?(zip)
-    zip != zip_code
+  def find_zip_code
+    @_find_zip_code ||= zip_code && ZipCode.find_by_zip(zip_code)
```

Comment:
it's a convention I've adopted for instance variables that should never be accessed from another method. Here's an example in Rails docs: http://guides.rubyonrails.org/action_controller_overview.html#accessing-the-session

It may be unnecessary, especially, since I'm trying to never access instance variables except through attr_reader, etc. Basically, that means that every instance variable that shouldn't have an attr_accessor starts with an underscore